lib/CsvImportAuditer.class.php

```diff
@@ -0,0 +1,308 @@
+                    'Warning! skipped row [%u/%u]: %s',
+                    $this->rowsAudited,
+                    $this->rowsTotal,
+                    $e->getMessage()
+                ));
+
+                continue;
+            }
+
+            ++$this->rowsAudited;
+            $this->log($this->progressUpdate($this->rowsAudited));
+        }
+
+        if (!empty($this->missingIds)) {
+            $this->log('');
+            $this->log('Source IDs not found in keymap data:');
+
+            foreach ($this->missingIds as $sourceId => $rowNumber) {
+                $this->log(sprintf('* %d (row %d)', $sourceId, $rowNumber));
+            }
+        }
+    }
+
+    public function loadCsvData($filename): League\Csv\ResultSet
+    {
+        $this->validateFileName($filename);
+
+        $reader = $this->readCsvFile($filename);
+        $stmt = new \League\Csv\Statement();
+        $records = $this->getRecords($reader, $stmt);
+
+        $this->rowsTotal = count($records);
+
+        return $records;
+    }
+
+    public function processRow($data): void
+    {
+        // Determine column name to check
+        $idColumnName = $this->getOption('idColumnName');
+
+        // Throw error if not ID value is found
+        if (empty($data[$idColumnName])) {
+            throw new UnexpectedValueException(sprintf('ID column %s not found', $idColumnName));
+        }
+
+        // Attempt to fetch keymap entry corresponding to source ID
+        $sourceId = trim($data[$idColumnName]);
+
+        if (false === $this->ormClasses['keymap']::getTargetId(
+            $this->getSourceName(),
+            $sourceId,
+            $this->getTargetName()
+        )) {
+            $this->missingIds[$sourceId] = $this->rowsAudited + 1;
+        }
+    }
+
+    public function progressUpdate($count): string
+    {
+        $freq = $this->getOption('progressFrequency');
+
+        if (1 == $freq) {
+            $msg = 'Row [%u/%u] audited';
+
+            $output = sprintf(
+                $msg,
+                $count,
+                $this->rowsTotal
+            );
+        } elseif ($freq > 1 && 0 == $count % $freq) {
+            $output = sprintf(
+                'Audited %u of %u rows...',
+                $count,
+                $this->rowsTotal
+            );
+        }
+
+        return $output;
+    }
+
+    public function getMissingIds(): array
+    {
+        return $this->missingIds;
+    }
+
+    //
+    // Protected methods
+    //
+
+    protected function log($msg): void
+    {
+        if (!$this->getOption('quiet')) {
+            echo $msg.PHP_EOL;
+        }
+    }
+
+    protected function logError($msg): void
+    {
+        // Write to error log (but not STDERR)
+        if (STDERR != $this->getErrorLogHandle()) {
+            fwrite($this->getErrorLogHandle(), $msg.PHP_EOL);
+        }
+    }
+
+    protected function getErrorLogHandle(): resource
+    {
+        if (null === $filename = $this->getOption('errorLog')) {
```

Comment:
Looks like this is expecting a `$filename` argument for this method that doesn't exist?

Reply:
It's getting the value, if set, from the "errorLog" option, if set.

Reply:
Oh right. :)